Add option to random load balance to >1 server nodes

kamran-m opened the pull request:
This closes #630 

I have also verified the change for a `client => server` with low number of client/server nodes, where the imbalance was more severe (note that here I used `peerConnectionCount`>= `numPeers=5`, which is the extreme case but it's enough to verify the solution):

<img width="1308" alt="screen shot 2017-06-16 at 10 51 25 pm" src="https://user-images.githubusercontent.com/3223733/27250495-65056392-52e6-11e7-9f1b-121ceeb6e3e3.png">

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/uber/tchannel-go?pullRequest=631) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/uber/tchannel-go?pullRequest=631) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/uber/tchannel-go?pullRequest=631) it.</sub>

prashantv commented on peer.go:
```diff
@@ -175,8 +186,8 @@ func (l *PeerList) Remove(hostPort string) error {
 func (l *PeerList) choosePeer(prevSelected map[string]struct{}, avoidHost bool) *Peer {
-	var psPopList []*peerScore
-	var ps *peerScore
+	var chosenPSList []*peerScore
```
presize chosenPSList to avoid a bunch of allocations on each selection, especially when it's > 1

prashantv commented on peer.go:
```diff
@@ -191,29 +202,43 @@ func (l *PeerList) choosePeer(prevSelected map[string]struct{}, avoidHost bool)
 }
 
+func randomSampling(psList []*peerScore) *peerScore {
+	peerRand := trand.NewSeeded()
```
please do not create a new rand for every single peer selection request.

prashantv commented on peer.go:
```diff
@@ -191,29 +202,43 @@ func (l *PeerList) choosePeer(prevSelected map[string]struct{}, avoidHost bool)
 	}
 
+	ps := randomSampling(chosenPSList)
 	if ps == nil {
```
why would we need this check?

kamran-m commented:
Thanks for the review @prashantv , I addressed your comments. I also added some benchmark tests with these results:
```
BenchmarkGetPeerWithPeerConnectionCount1  	 5000000	       348 ns/op
BenchmarkGetPeerWithPeerConnectionCount10 	 1000000	      1490 ns/op
```
and also used the same test in the old code  to compare with `peerConnectionCount = 1` case above:
```
BenchmarkGetPeerOld-8   	 5000000	       267 ns/op
```
Please let me know if the `267 => 348` is of any concern.

prashantv commented on peer_test.go:
```diff
@@ -697,6 +697,118 @@ func TestPeerSelectionRanking(t *testing.T) {
+
+	for i := 0; i < b.N; i++ {
+		peer, _ := ch.Peers().Get(nil)
+		if peer == nil {
```
should this ever happen? If not, maybe we should do a `t.Fatal` instead of a println

kamran-m replied:
It shouldn't, I just added it to guard against compiler optimization not to artificially lower the runtime of the benchmark. Changed it to Fatal!

kamran-m replied:
It shouldn't, I just added it to guard against compiler optimization not to artificially lower the runtime of the benchmark. Changed it to Fatal!

prashantv commented on peer_test.go:
```diff
@@ -697,6 +697,118 @@ func TestPeerSelectionRanking(t *testing.T) {
+
+func BenchmarkGetPeerWithPeerConnectionCount10(b *testing.B) {
+	numPeers := 10
+	peerConnectionCount := uint32(10)
```
Instead of copying + pasting the whole benchmark, abstract common logic out into a function, and then call it with different args for the `peerConnectionCount` and maybe even `numPeers`

kamran-m replied:
done!

prashantv commented on peer_test.go:
```diff
@@ -697,6 +697,118 @@ func TestPeerSelectionRanking(t *testing.T) {
+	}{
+		// the higher `peerConnectionCount` is, the smoother the impact of uneven scores
+		// become as we are random sampling among `peerConnectionCount` peers
+		{numPeers: 10, peerConnectionCount: 1, distMin: 1000, distMax: 1000},
```
can you add `peerConnectionCount: 2`, i imagine this will be a pretty small value normally?

kamran-m replied:
Indeed as the impact of it get's smaller with every extra connection. Added the case for `2`!

kamran-m replied:
Indeed as the impact of it get's smaller with every extra connection. Added the case for `2`!

prashantv commented on peer_test.go:
```diff
@@ -697,6 +697,118 @@ func TestPeerSelectionRanking(t *testing.T) {
+	for _, tc := range testCases {
+		// Selected is a map from rank -> [peer, count]
+		// It tracks how often a peer gets selected at a specific rank.
+		selected := make([]map[string]int, tc.numPeers)
```
why do you make a slice here, why can't we just create a singel map in the loop for `numIterations`, and do the `testDistribution` call right there too? Makes 3 loops 1 loop, removes a slice, and simplifies the test a little.

kamran-m replied:
I am not sure if I follow. This is a similar test as `TestPeerSelectionRanking` where we are checking the distribution of the ranking after `numIterations` which means we have to do the checking outside of the loop. Or am I missing something here?

kamran-m replied:
I am not sure if I follow. This is a similar test as `TestPeerSelectionRanking` where we are checking the distribution of the ranking after `numIterations` which means we have to do the checking outside of the loop. Or am I missing something here?

prashantv commented on peer_test.go:
```diff
@@ -697,6 +697,118 @@ func TestPeerSelectionRanking(t *testing.T) {
+	b.ResetTimer()
+
+	for i := 0; i < b.N; i++ {
+		peer, _ := ch.Peers().Get(nil)
```
we should think about testing the impact of passing a non-nil blacklist with `peerConnectionCount` > 1

kamran-m replied:
You mean if `prevSelected` actually has some selected peers? I am assuming if a client uses `peerConnectionCount > 1`, it should keep it into account that the prevSelected peers is going to be a bigger list. Or do you have another specific issue in mind?

kamran-m replied:
You mean if `prevSelected` actually has some selected peers? I am assuming if a client uses `peerConnectionCount > 1`, it should keep it into account that the prevSelected peers is going to be a bigger list. Or do you have another specific issue in mind?

kamran-m commented on peer.go:
```diff
@@ -48,6 +47,9 @@ var (
 	ErrNoNewPeers = errors.New("no new peer available")
 
+	// ErrZeroPeerConnectionCount indicates that the peer connection count is set to zero.
+	ErrZeroPeerConnectionCount = errors.New("peer connection count must be greater than 0")
```
Makes sense!

kamran-m commented:
I agree with the idea that the core should not be strategy aware. As I commented on #630 I think all the logic regarding peer connection count, score strategy, peer heap, etc should move to something like PeerSelectionStrategy. 

I am not sure why you think with 2 connected peers we won't need this core change and also get a better load distribution. This change is basically letting the client decide how many connected peers they want. So in cases, like ours, were we have a very low number of server/client nodes, it might actually be helpful to use 5 connected peers as you can verify with the stats I posted above.

Let me know your thoughts.
